server/extension/extension.go

```diff
@@ -277,81 +413,107 @@ func (m *Manager) registerExtensions(r *mux.Router, extConfigs *ExtensionConfigs
 	return nil
 }
 
+// authorize will enforce rbac rules are satified for the given RequestResources.
+// The following validations are executed:
+//   - enforce the subject has permission to read application/project provided
+//     in HeaderArgoCDApplicationName and HeaderArgoCDProjectName.
+//   - enforce the subject has permission to invoke the extension identified by
+//     extName.
+//   - enforce that all resources provided in the HeaderArgoCDResourceGVKName belong
+//     to the application identified by HeaderArgoCDApplicationName.
+//
+// If all validations are satified it will return the Application resource
+func (m *Manager) authorize(ctx context.Context, rr *RequestResources, extName string) (*v1alpha1.Application, error) {
+	if m.rbac == nil {
+		return nil, fmt.Errorf("rbac enforcer not set in extension manager")
+	}
+	appRBACName := security.AppRBACName(rr.ApplicationNamespace, rr.ProjectName, rr.ApplicationNamespace, rr.ApplicationName)
+	if err := m.rbac.EnforceErr(ctx.Value("claims"), rbacpolicy.ResourceApplications, rbacpolicy.ActionGet, appRBACName); err != nil {
+		return nil, fmt.Errorf("application authorization error: %s", err)
+	}
+
+	if err := m.rbac.EnforceErr(ctx.Value("claims"), rbacpolicy.ResourceExtensions, rbacpolicy.ActionInvoke, extName); err != nil {
+		return nil, fmt.Errorf("unauthorized to invoke extension %q: %s", extName, err)
+	}
+
+	// just retrieve the app after checking if subject has access to it
+	app, err := m.application.Get(rr.ApplicationNamespace, rr.ApplicationName)
```

Comment:
We also need to check the returned `app` spec to make sure that the `spec.project` field matches the provided `rr.ProjectName`.

If it does not, we need to return a generic error to avoid revealing that the application exists (the user hasn't proven yet that they're allowed to know that).

Reply:
Good catch.
Added: `project mismatch provided in the header`